pkg/networkserver/downlink.go

```diff
@@ -144,15 +144,20 @@ func (ns *NetworkServer) generateDownlink(ctx context.Context, dev *ttnpb.EndDev
 	}
 	cmds = append(cmds, dev.MACState.PendingRequests...)
 
+	mType := ttnpb.MType_UNCONFIRMED_DOWN
 	cmdBuf := make([]byte, 0, maxDownLen)
 	for _, cmd := range cmds {
+		logger := logger.WithField("cid", cmd.CID)
+		logger.Debug("Adding MAC command to buffer...")
 		var err error
-		logger.WithField("cid", cmd.CID).Debug("Adding MAC command to buffer...")
 		cmdBuf, err = spec.AppendDownlink(cmdBuf, *cmd)
-
 		if err != nil {
 			return nil, errEncodeMAC.WithCause(err)
 		}
+		if mType == ttnpb.MType_UNCONFIRMED_DOWN && spec[cmd.CID].ExpectAnswer && dev.MACState.DeviceClass == ttnpb.CLASS_C {
```

Comment:
Shouldn't we also do this for class B?

Reply:
The reason to make the downlink confirmed is for the device to respond within the class C timeout (V3 default 15 seconds). If it responds later, NS keeps sending MAC requests.

In class B, we need to make them confirmed as well, if the device does not respond before the next ping slot. LoRaWAN 1.0.3 does not specify answering class B frames; it only states that they're handled like class A, and in class A, the answer may come later than the next ping slot.

So I'm fine confirmed downlink of MAC commands where expect an answer in class B as well, but this is not based on device behavior but rather what's not specified.

Reply:
Is this something we need to raise in the TC as well?

Reply:
Added reference in #19 so that we don't forget. If I have time I'll raise a CR for Berlin AMM.